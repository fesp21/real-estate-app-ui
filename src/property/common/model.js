import { Model, fk } from "redux-orm";

export class Property extends Model {
  static modelName = "Property";
  static options() {
    return {
      idAttribute: "_id"
    };
  }

  static fields = {
    user: fk("User")
  };
}
