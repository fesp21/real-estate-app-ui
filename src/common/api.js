export async function fetchAPI(
  url,
  method = "GET",
  body = null,
  isBlob = false
) {
  let request;

  if (method === "POST") {
    request = fetch(url, {
      method,
      body: isBlob ? body : JSON.stringify(body),
      headers: {
        "Accept": "application/json",
        "Content-Type": isBlob ? "multipart/form-data" : "application/json"
      }
    });
  } else {
    request = fetch(url, {
      method
    });
  }

  return request
    .then(response => response.json().then(json => ({
      status: response.status,
      statusType: response.statusType,
      json
    })))
    .then(({ status, statusType, json }) => {
      if (status !== 200 || !json.success) {
        const unknownError = json.errors
          ? json.errors
          : `Unknown Error. Status Type : ${statusType}`;
        const errorMsg = json.message ? json.message : unknownError;
        return Promise.reject(errorMsg);
      }
      return json;
    })
    .catch(e => {
      return Promise.reject(`Unknown Error : ${e}`);
    });
}

export async function xhrAPI(
  url,
  body,
) {

  const xhr = new XMLHttpRequest();
  xhr.open('POST', url);
  xhr.send(body);
  xhr.onreadystatechange = (e) => {
    console.log('onReady');
    if (xhr.status === 200) {
      console.log('xhr.status',xhr);
      return JSON.parse(xhr.responseText);
    } else {
      console.log('xhr.reject',xhr);
      return Promise.reject(xhr.response)
    }
  };
  console.log('ended');
  return xhr;

}



