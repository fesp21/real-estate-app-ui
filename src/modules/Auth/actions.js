export const ACTION_TYPES = {
  LOGIN_REQUEST : 'LOGIN_REQUEST',
  LOGIN_SUCCESS : 'LOGIN_SUCCESS',
  LOGIN_FAILURE : 'LOGIN_FAILURE',
  REGISTER_REQUEST : 'REGISTER_REQUEST',
  REGISTER_SUCCESS : 'REGISTER_SUCCESS',
  REGISTER_FAILURE : 'REGISTER_FAILURE',
  LOGOUT : 'LOGOUT',
};

function login(credentials,redirectUrl) {
  return {
    type: ACTION_TYPES.LOGIN_REQUEST,
    credentials,
    redirectUrl
  }
}

function logout() {
  return {
    type: ACTION_TYPES.LOGOUT
  }
}

function register(params) {
  return {
    type: ACTION_TYPES.REGISTER_REQUEST,
    params
  }
}

export const ACTIONS = {
  login,
  register,
  logout
};