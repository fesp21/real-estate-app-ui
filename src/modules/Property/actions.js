import includes from 'lodash/includes';

export const ACTION_TYPES = {
  PROPERTY_REQUEST : 'PROPERTY_REQUEST',
  PROPERTY_SUCCESS : 'PROPERTY_SUCCESS',
  PROPERTY_FAILURE : 'PROPERTY_FAILURE',
  // CATEGORY_REQUEST : 'CATEGORY_REQUEST',
  // CATEGORY_SUCCESS : 'CATEGORY_SUCCESS',
  // CATEGORY_FAILURE : 'CATEGORY_FAILURE',
  FILTER_CHANGE : 'FILTER_CHANGE',
  PROPERTY_RESET : 'PROPERTY_RESET',
  PROPERTY_FAVORITE_REQUEST : 'PROPERTY_FAVORITE_REQUEST',
  PROPERTY_FAVORITE_SUCCESS : 'PROPERTY_FAVORITE_SUCCESS',
  PROPERTY_FAVORITE_FAILURE : 'PROPERTY_FAVORITE_FAILURE',
  FAVORITE_SUCCESS : 'FAVORITE_SUCCESS',
  FAVORITE_REQUEST : 'FAVORITE_REQUEST',
  FAVORITE_FAILURE : 'FAVORITE_FAILURE',
  LISTING_CHANGE : 'LISTING_CHANGE',
  PROPERTY_SAVE_REQUEST: 'PROPERTY_SAVE_REQUEST',
  PROPERTY_SAVE_SUCCESS: 'PROPERTY_SAVE_SUCCESS',
  PROPERTY_SAVE_FAILURE: 'PROPERTY_SAVE_FAILURE',
};

function fetchProperties(params) {
  return {
    type: ACTION_TYPES.PROPERTY_REQUEST,
    params:params
  }
}

function fetchFavorites(params ='') {
  return {
    type: ACTION_TYPES.FAVORITE_REQUEST,
    params
  }
}


// function fetchCategories(params) {
//   return {
//     type: ACTION_TYPES.CATEGORY_REQUEST,
//     params
//   }
// }

function changeFormValue(field,value) {
  return {
    type: ACTION_TYPES.FILTER_CHANGE,
    field,
    value
  }
}

function invalidateProperty() {
  return {
    type: ACTION_TYPES.PROPERTY_RESET,
  }
}

function favoriteProperty(property) {
  return {
    type:ACTION_TYPES.PROPERTY_FAVORITE_REQUEST,
    params:{
      itemID:property._id,
      newItemAttributes: {
        isFavorited : !property.isFavorited
      }
    },
  }
}

function changeListingValue(payload) {
  return {
    type:ACTION_TYPES.LISTING_CHANGE,
    payload,
  }
}

function saveProperty() {
  return {
    type:ACTION_TYPES.PROPERTY_SAVE_REQUEST
  }
}

export const ACTIONS = {
  fetchProperties,
  // fetchCategories,
  changeFormValue,
  invalidateProperty,
  favoriteProperty,
  fetchFavorites,
  changeListingValue,
  saveProperty
};