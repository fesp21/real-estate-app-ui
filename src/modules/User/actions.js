export const ACTION_TYPES = {
  USER_REQUEST : 'USER_REQUEST',
  USER_SUCCESS : 'USER_SUCCESS',
  USER_FAILURE : 'USER_FAILURE',
};

function fetchUser(params) {
  return {
    type: ACTION_TYPES.USER_REQUEST,
    params:params
  }
}
export const ACTIONS = {
  fetchUser,
};