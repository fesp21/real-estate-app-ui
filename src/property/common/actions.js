export const ACTION_TYPES = {
  PROPERTY_REQUEST: 'PROPERTY_REQUEST',
  PROPERTY_SUCCESS: 'PROPERTY_SUCCESS',
  PROPERTY_FAILURE: 'PROPERTY_FAILURE',
  PROPERTY_RESET: 'PROPERTY_RESET',
  PROPERTY_FAVORITE_REQUEST: 'PROPERTY_FAVORITE_REQUEST',
  PROPERTY_FAVORITE_SUCCESS: 'PROPERTY_FAVORITE_SUCCESS',
  PROPERTY_FAVORITE_FAILURE: 'PROPERTY_FAVORITE_FAILURE',
  FAVORITES_SUCCESS: 'FAVORITES_SUCCESS',
  FAVORITES_REQUEST: 'FAVORITES_REQUEST',
  FAVORITES_FAILURE: 'FAVORITES_FAILURE',
  PROPERTY_SAVE_REQUEST: 'PROPERTY_SAVE_REQUEST',
  PROPERTY_SAVE_SUCCESS: 'PROPERTY_SAVE_SUCCESS',
  PROPERTY_SAVE_FAILURE: 'PROPERTY_SAVE_FAILURE',
  FILTER_CHANGE: 'FILTER_CHANGE',
  LISTING_UPDATE_ITEM: 'LISTING_UPDATE_ITEM',
  FILTER_RESET: 'FILTER_RESET',
  PROPERTY_FAVORITE_OPTIMISTIC_UPDATE: 'PROPERTY_FAVORITE_OPTIMISTIC_UPDATE',
};

function fetchProperties(params) {
  return {
    type: ACTION_TYPES.PROPERTY_REQUEST,
    params,
  };
}

function fetchFavorites(params) {
  return {
    type: ACTION_TYPES.FAVORITES_REQUEST,
    params,
  };
}

function changeFormValue(field, value) {
  return {
    type: ACTION_TYPES.FILTER_CHANGE,
    field,
    value,
  };
}

function invalidateProperty() {
  return {
    type: ACTION_TYPES.PROPERTY_RESET,
  };
}

function favoriteProperty(property) {
  return {
    type: ACTION_TYPES.PROPERTY_FAVORITE_REQUEST,
    params: {
      itemID: property._id,
      newItemAttributes: {
        isFavorited: !property.isFavorited,
      },
    },
  };
}

function changeListingValue(payload) {
  return {
    type: ACTION_TYPES.LISTING_UPDATE_ITEM,
    payload,
  };
}

function resetFilter() {
  return {
    type: ACTION_TYPES.FILTER_RESET,
  };
}

function saveProperty() {
  return {
    type: ACTION_TYPES.PROPERTY_SAVE_REQUEST,
  };
}

export const ACTIONS = {
  fetchProperties,
  changeFormValue,
  invalidateProperty,
  favoriteProperty,
  fetchFavorites,
  changeListingValue,
  saveProperty,
  resetFilter,
};
