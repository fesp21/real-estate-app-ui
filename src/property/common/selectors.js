import orm from '../../common/orm';
import { createSelector } from 'reselect';
import { createSelector as ormSelector } from 'redux-orm';

const propertyResults = state => state.propertyReducer.results;
const propertyIsFetching = state => state.propertyReducer.isFetching;
const propertyCategories = state => state.propertyReducer.categories;
const propertyTypes = state => state.propertyReducer.types;
const propertyAmenities = state => state.propertyReducer.amenities;
const propertyFilters = state => state.propertyReducer.filters;
const propertyListings = state => state.propertyReducer.listings;
const ormReducer = state => state.ormReducer;
const getPropertyID = (state, props) => props.property._id;

const filterResults = ({ Property, User }, results) => results.map((id) => {
  const property = Property.withId(id).ref;
  return Object.assign({}, property, { user: User.withId(property.user).ref });
});

const fetchProperties = createSelector(
  ormReducer,
  propertyResults,
  ormSelector(orm, (ormSession, results) => filterResults(ormSession, results)),
);

const fetchFavorites = createSelector(
  ormReducer,
  ormSelector(orm, ({ Property, User }) => Property.all().toRefArray().filter(property => property.isFavorited).map(property => Object.assign({}, property, { user: User.withId(property.user).ref }))),
);

const fetchProperty = createSelector(
  ormReducer,
  getPropertyID,
  ormSelector(orm, ({ Property, User }, id) => {
    const property = Property.withId(id).ref;
    return Object.assign({}, property, { user: User.withId(property.user).ref });
  }),
);

const getCategoriesWithAny = createSelector(
  propertyCategories,
  categories => categories.concat('Any').reverse(),
);

const getCategories = createSelector(
  propertyCategories,
  categories => categories,
);

const isFetching = createSelector(
  propertyIsFetching,
  isFetching => isFetching,
);

const getFilters = createSelector(
  propertyFilters,
  filters => filters,
);

const getListing = createSelector(
  propertyListings,
  listings => listings,
);

const fetchComments = createSelector(
  [],
  comments => [],
);

const getTypes = createSelector(
  propertyTypes,
  types => types,
);

const getAmenities = createSelector(
  propertyAmenities,
  amenities => amenities,
);

export const SELECTORS = {
  isFetching,
  fetchProperties,
  fetchComments,
  fetchFavorites,
  fetchProperty,
  getCategories,
  getFilters,
  getListing,
  getCategoriesWithAny,
  getTypes,
  getAmenities,
};
