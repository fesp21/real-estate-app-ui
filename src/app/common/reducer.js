import { ACTION_TYPES } from "./actions";

export const COUNTRY_KEY = "COUNTRY";
const DEFAULT_COUNTRY = "Kuwait";

// // reducer
const initialState = {
  bootstrapped: true,
  country: DEFAULT_COUNTRY
};

export default function appReducer(state = initialState, action = {}) {
  switch (action.type) {
    case ACTION_TYPES.BOOT_REQUEST:
      return { ...state, bootstrapped: false };
    case ACTION_TYPES.BOOT_SUCCESS:
      return { ...state, bootstrapped: true };
    case ACTION_TYPES.COUNTRY_CHANGED:
      return { ...state, country: action.country };
    default:
      return state;
  }
}
