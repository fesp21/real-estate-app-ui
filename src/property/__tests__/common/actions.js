import { ACTION_TYPES, ACTIONS } from '../../common/actions';

describe('App Component Actions', () => {

  test('fetchProperties action', () => {
    expect(ACTIONS.fetchProperties({
      id:1
    })).toMatchSnapshot();
  });

  test('fetchFavorites action', () => {
    expect(ACTIONS.fetchFavorites({
      id:1
    })).toMatchSnapshot();
  });

  test('changeFormValue action', () => {
    expect(ACTIONS.changeFormValue('bedroom',1)).toMatchSnapshot();
  });

  test('invalidateProperty action', () => {
    expect(ACTIONS.invalidateProperty()).toMatchSnapshot();
  });

  test('changeListingValue action', () => {
    expect(ACTIONS.changeListingValue('attributes',{
      title:'property title',
      address: {
        city: 'Kuwait City',
        state: 'Kuwait City',
        country: 'Kuwait',
        latitude: 29.3667,
        longitude: 47.9667,
      },
      meta: {
        bedroom: 'Studio',
        bathroom: '1',
        kitchen: '1',
        area: '220.5',
        parking: '1',
      },
      images: [],
      tags: ['New', 'Duplex'],
      amenities: ['Swimming Pool'],
    })).toMatchSnapshot();
  });

  test('reset filter', () => {
    expect(ACTIONS.resetFilter()).toMatchSnapshot();
  });

  test('reset filter', () => {
    expect(ACTIONS.saveProperty()).toMatchSnapshot();
  });

});