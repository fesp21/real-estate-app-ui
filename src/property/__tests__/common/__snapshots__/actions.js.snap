exports[`App Component Actions changeFormValue action 1`] = `
Object {
  "field": "bedroom",
  "type": "FILTER_CHANGE",
  "value": 1,
}
`;

exports[`App Component Actions changeListingValue action 1`] = `
Object {
  "payload": "attributes",
  "type": "LISTING_CHANGE",
}
`;

exports[`App Component Actions fetchFavorites action 1`] = `
Object {
  "params": Object {
    "id": 1,
  },
  "type": "FAVORITES_REQUEST",
}
`;

exports[`App Component Actions fetchProperties action 1`] = `
Object {
  "params": Object {
    "id": 1,
  },
  "type": "PROPERTY_REQUEST",
}
`;

exports[`App Component Actions invalidateProperty action 1`] = `
Object {
  "type": "PROPERTY_RESET",
}
`;

exports[`App Component Actions reset filter 1`] = `
Object {
  "type": "FILTER_RESET",
}
`;

exports[`App Component Actions reset filter 2`] = `
Object {
  "type": "PROPERTY_SAVE_REQUEST",
}
`;
